### PR TITLE
HPCC-16683 Avoid unnecessary traversing the body of delayed calls

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review

File: system/jlib/jarray.hpp [hunk 1]
```diff
@@ -67,9 +66,9 @@ class jlib_decl Allocator
     void doSwapWith(Allocator & other);
 
 protected:
-    void * _head;
-    aindex_t used;
-    aindex_t max;
+    void * _head = nullptr;
```
Comment (ghalliday) [inline system/jlib/jarray.hpp]: This change was needed to work around a problem with the order that constructors for static objects were called.

File: ecl/hql/hqlexpr.cpp [hunk 2]
```diff
@@ -10476,10 +10476,13 @@ IHqlExpression *CHqlExternalDatasetCall::clone(HqlExprArray &newkids)
 
 //==============================================================================================================
 
-CHqlDelayedCall::CHqlDelayedCall(IHqlExpression * _param, ITypeInfo * _type, HqlExprArray &_ownedOperands) : CHqlExpressionWithType(no_call, _type, _ownedOperands), param(_param)
+CHqlDelayedCall::CHqlDelayedCall(IHqlExpression * _funcdef, ITypeInfo * _type, HqlExprArray &_ownedOperands) : CHqlExpressionWithType(no_call, _type, _ownedOperands), funcdef(_funcdef)
 {
-    infoFlags |= (param->getInfoFlags() & HEFalwaysInherit);
-    infoFlags2 |= (param->getInfoFlags2() & HEF2alwaysInherit);
+    IHqlExpression * body = funcdef;
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: This is the main change - the rest are renaming a field, and improving the logging of transform stats when enabled.

**Comment by richardkchapman** [comment]
Typo in commit title

**Comment by ghalliday** [comment]
typo fixed.

**Comment by ghalliday** [comment]
@richardkchapman @shamser I am wondering whether this should go into 6.2.2 since it that would save developer time sooner.
I have several other optimizations coming soon which should also be considered for 6.2.2.

**Comment by ghalliday** [comment]
@shamser now retargeted to 6.2.2

File: ecl/hql/hqlexpr.ipp [hunk 3]
```diff
@@ -988,15 +988,15 @@ typedef class MapXToMyClassViaBase<IAtom *, IAtom *, IFileContents, IFileContent
...
-    virtual IIdAtom * queryId() const { return param->queryId(); }
+    virtual IAtom * queryName() const { return funcdef->queryName(); }
+    virtual IIdAtom * queryId() const { return funcdef->queryId(); }
     virtual void sethash();
     virtual bool equals(const IHqlExpression & other) const;
     virtual IHqlExpression *clone(HqlExprArray &newkids);
-    virtual IHqlExpression *queryFunctionDefinition() const { return param; };
+    virtual IHqlExpression *queryFunctionDefinition() const { return funcdef; };
 public:
     static IHqlExpression *makeDelayedCall(IHqlExpression * _param, HqlExprArray &operands);
```
Comment (shamser) [inline ecl/hql/hqlexpr.ipp]: @ghalliday Should the first argument be named _funcdef?

File: ecl/hql/hqlexpr.ipp [hunk 4]
```diff
@@ -988,15 +988,15 @@ typedef class MapXToMyClassViaBase<IAtom *, IAtom *, IFileContents, IFileContent
 
 class CHqlDelayedCall: public CHqlExpressionWithType
 {
-    OwnedHqlExpr param;
+    OwnedHqlExpr funcdef;
 protected:
     CHqlDelayedCall(IHqlExpression * _param, ITypeInfo * type, HqlExprArray &parms);
```
Comment (shamser) [inline ecl/hql/hqlexpr.ipp]: @ghalliday Should the first argument be named _funcdef?  Also, note that CHqlDelayedCall drerived classes calls CHqlDelayedCall constructor with first parameter named _param - should these be renamed too?

**Review by shamser — COMMENTED** [review]
@ghalliday Couple of comments relating to consistency.

**Comment by ghalliday** [comment]
@richardkchapman parameters renamed.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed.  Looks good.

**Pull request merged**